datafusion/physical-expr/src/expressions/case.rs

```diff
@@ -122,6 +123,276 @@ fn is_cheap_and_infallible(expr: &Arc<dyn PhysicalExpr>) -> bool {
+    /// └─────────┘     └─────────┘└───────────┘                            └─────────┘└───────────┘
+    /// row indices
+    ///                   partial     partial                                 partial     partial
+    ///                   indices     results                                 indices     results
+    /// ```
+    fn add_branch_result(
+        &mut self,
+        row_indices: &ArrayRef,
+        value: ColumnarValue,
+    ) -> Result<()> {
+        match value {
+            ColumnarValue::Array(a) => {
+                assert_eq!(a.len(), row_indices.len());
+                if row_indices.len() == self.partial_result_indices.len() {
+                    self.set_covering_result(ColumnarValue::Array(a));
+                } else {
+                    self.add_partial_result(row_indices, a.to_data());
+                }
+            }
+            ColumnarValue::Scalar(s) => {
+                if row_indices.len() == self.partial_result_indices.len() {
+                    self.set_covering_result(ColumnarValue::Scalar(s));
+                } else {
+                    self.add_partial_result(
+                        row_indices,
+                        s.to_array_of_size(row_indices.len())?.to_data(),
+                    );
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Adds a partial result array.
+    ///
+    /// This method adds the given array data as a partial result and updates the index mapping
+    /// to indicate that the specified rows should take their values from this array.
+    /// The partial results will be merged into a single array when finish() is called.
+    fn add_partial_result(&mut self, row_indices: &ArrayRef, row_values: ArrayData) {
+        // Covering results and partial results are mutually exclusive.
+        // We can assert this since the case evaluation methods are written to only evaluate
+        // each row of the record batch once.
+        assert!(self.covering_result.is_none());
+
+        self.partial_results.push(row_values);
+        let array_index = self.partial_results.len();
+
+        for row_ix in row_indices.as_primitive::<UInt32Type>().values().iter() {
+            self.partial_result_indices[*row_ix as usize] = array_index;
+        }
+    }
+
+    /// Sets a covering result that applies to all rows.
+    ///
+    /// This is an optimization for cases where all rows evaluate to the same result.
+    /// When a covering result is set, the builder will return it directly from finish()
+    /// without any merging overhead.
+    fn set_covering_result(&mut self, value: ColumnarValue) {
+        // Covering results and partial results are mutually exclusive.
+        // We can assert this since the case evaluation methods are written to only evaluate
+        // each row of the record batch once.
+        assert!(self.partial_results.is_empty());
```

Comment:
if you used the enum above you could avoid this assert (the compiler would ensure it for you)